### PR TITLE
Add a query interface as well as indexes.

### PR DESCRIPTION
**Pull request by ekr-cfa**

### CODE REVIEW

**Comment by ekr-cfa** [comment]
This update replaces the macro `people_query!` with a new API that allow you to query with a set of tuples, like so:

```
let people = context.query_people(((Age, 42), (RiskCategoryType, RiskCategory::High)));
```

The trick here is that `query_people()` takes an implementation of `Query`

```
    fn query_people<T: Query>(&self, q: T) -> Vec<PersonId>;
```

And then we have implementations of `Query` for tuples up to length 20.

**Review by k88hudson-cfa — APPROVED** [review]
LGTM

File: src/people.rs [hunk 1]
```diff
@@ -479,11 +689,190 @@ impl ContextPeopleExt for Context {
...
+        // 1. Walk through each property and collect the index entry
+        // corresponding to the value.
+        for (t, hash) in property_hashes {
+            let mut index = data_container.get_index_ref(t).unwrap();
+            index.index_unindexed_people(self);
+
+            // Update the index.
+            if let Some(lookup) = &index.borrow_mut().lookup {
+                if let Some(matching_people) = lookup.get(&hash) {
+                    indexes.push(matching_people.clone()); // UGH
```
[the remark this answers is not in the document]
Comment (ekr-cfa) [reply src/people.rs:824]: I see you fixed this in your branch. Why don't you out that up for a PR and I can merge it in.

File: src/people.rs [hunk 2]
```diff
@@ -3,14 +3,186 @@ use crate::{
...
+    }
+}
+
+impl Hasher for IndexValueHasher {
+    fn write(&mut self, bytes: &[u8]) {
+        self.buf.extend_from_slice(bytes);
+    }
+
+    fn finish(&self) -> u64 {
+        panic!("Unimplemented")
```
[the remark this answers is not in the document]
Comment (ekr-cfa) [reply src/people.rs:122]: Yes, I didn't know about this.

**Pull request merged**